base/inference.jl

```diff
@@ -4109,6 +4163,7 @@ function inlining_pass(e::Expr, sv::InferenceState)
                         res = inlining_pass(e, sv)
                     else
                         e.args = Any[GlobalRef(Main.Base,:*), Expr(:call, GlobalRef(Main.Base,:*), a1, a1), a1]
+                        e.args[2].typ = e.typ
                         res = inlining_pass(e, sv)
```

Comment:
Somehow the call to `inlining_pass` causes `e.args[2].typ` to become annotated with the correct type, even without your change here. I assume this was a happy accident, but how does it happen?